Fixed backward compatibility for `send_email_when_published`

allouis opened the pull request:
no-issue 

refs #12343 

This restores the functionality of `send_email_when_published` query param to the Admin API's Post Edit endpoint, as well as restoring the field in both the Content & Admin API, and patching in support for the `fields` param.

Also includes a migration to drop the old column, now that it's no longer used.

allouis committed:
Handled send_email_when_published in Posts API

no-issue

This restores backwards compatibility of the Posts API allowing existing
clients to continue to use the `send_email_when_published` flag. This
change uses two edits, which is unfortunate. The reason being is that
this is an API compatibility issue, not a model issue, so we shouldn't
introduce code to the model layer to handle it. The visibility property
of the model is used to determine how to fall back, and because it can
be left out of the API request, and relies on a default in the settings,
we require that the model decide on the `visibility` before we run our
fallback logic (or we duplicate the `visibility` default at the cost of
maintainence in the future)

allouis committed:
Dropped send_email_when_published column from posts

no-issue

Since this column is not used anymore, we can drop it from the table.
The down migration includes repopulating the data, though, because we
require a DDL command first, we are unable to use the transaction
created by knex-migrator, so we create our own transaction and handle
commiting and rolling it back.

allouis committed:
Updated importer to handle send_email_when_published

no-issue

Because we currently export this value from Ghost, we should correctly
import it. This follows the same logic as the migrations for this value.

allouis committed:
Included send_email_when_published in API response

no-issue

As our v3 API documentation includes `send_email_when_published` we must
retain backward compatibility by calculating the property.

allouis committed:
Fixed fields filter with send_email_when_published

allouis commented:
@daniellockyer @rishabhgrg @kevinansfield The migration needs some :eyes: https://github.com/TryGhost/Ghost/pull/12357/commits/0d932b4c19072566d82d0e692484d4c7cf0ecda2

allouis commented on core/server/data/migrations/versions/3.38/07-remove-send-email-when-published-column.js:
```diff
@@ -0,0 +1,43 @@
+            defaultTo: false
+        });
+
+        const trx = await connection.transaction();
```
Maybe this should be a separate migration? :thinking:

kevinansfield replied:
We've always kept DDL queries separate from DML before. I suppose having it in a separate migration allows for separating db state between migrations but not sure how much that actually buys us 🤔

kevinansfield replied:
This is only in the `down` command. If we perform a rollback and there's an error when migrating down do we then do a rollforward to get back to the last "good" state?

allouis replied:
Hmmm, a roll forward would completely drop the column though, which is the last "good" state - it's a weird one!

kevinansfield replied:
I think I'd lean more towards following existing patterns and keeping the migrations separate until we have a bigger discussion around the implications?

From the `knex-migrator` usage docs:
> Don't mix DDL/DML statements in a migration script. In MySQL DDL statements use implicit commits.

Keeping them separate also reduces cognitive overhead when deciding what to put into a migration 🤔 If there's no good reason any more to keep them separate then we should think about updating the usage docs and making sure everyone knows when it's ok to mix DDL/DML.

allouis replied:
Cool - sounds good to me :relaxed:

allouis replied:
I've updated this in the latest here: https://github.com/TryGhost/Ghost/pull/12357/commits/2bb89e9dae655412cdf97e1e81feb5a707443bdd

allouis committed:
Added safety checks to frame properties

no-issue

Some parts of the code pass a manually created "frame" which is missing
lots of properties, so we check for the existence of all of them before
using them.

allouis commented:
Hey @daniellockyer I'm not sure why the migrations test is failing :thinking: I can't replicate locally with either MySQL or SQLite!

The error seems to point to a missing column which is weird :raised_eyebrow: 
```
update `posts` set `send_email_when_published` = true where not `email_recipient_filter` = 'none' - SQLITE_ERROR: no such column: email_recipient_filter
```

daniellockyer commented:
@allouis Would you be able to rebase on top of `master`? I'm also investigating this elsewhere, but my latest commit should fix it for now 🙂

allouis committed:
Fixed 3.1 migration to include columnDefinition

no-issue

We require that migrations have all the information they need contained
within them as they run in an unknown state of the codebase, which could
be from the commit they are introduced, to any future commit. In this
case the column definition is removed from the schema in 3.38 and the
migration would fail when run in this version or later.

allouis committed:
fixup! Dropped send_email_when_published column from posts
